# Project - Guide school to activities to improve G3 grades

![Data Science Workflow](img/ds-workflow.png)

## Goal of Project
- Explore the dataset from lesson further
- Follow the Data Science process to understand it better
- It will be your task to identify possible activies to improve G3 grades
- NOTE: We have very limited skills, hence, we must limit our ambitions in our analysis

## Step 1: Acquire
- Explore problem
- Identify data
- Import data

### Step 1.a: Import libraries
- Execute the cell below (SHIFT + ENTER)

In [1]:
import pandas as pd

### Step 1.b: Read the data
- Use ```pd.read_csv()``` to read the file `files/student-mat.csv`
- NOTE: Remember to assign the result to a variable (e.g., ```data```)

In [2]:
data = pd.read_csv('files/student-mat.csv')

### Step 1.c: Inspect the data
- Call ```.head()``` on the data to see all is as expected

In [3]:
data.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]

### Step 1.d: Check length of data
- Call ```len(...)``` on the data
- Result: There should be 395 rows of data

In [5]:
len(data)

395

## Step 2: Prepare
- Explore data
- Visualize ideas
- Cleaning data

### Notice
- We will not cover visualization in this lecture
- We also know, that the data is clean - but we will do validations here anyway

### Step 2.a: Check the data types
- This step tells you if some numeric column is not represented numeric.
- Get the data types by ```.dtypes```

In [6]:
data.dtypes

school        object
sex           object
age            int64
address       object
famsize       object
Pstatus       object
Medu           int64
Fedu           int64
Mjob          object
Fjob          object
reason        object
guardian      object
traveltime     int64
studytime      int64
failures       int64
schoolsup     object
famsup        object
paid          object
activities    object
nursery       object
higher        object
internet      object
romantic      object
famrel         int64
freetime       int64
goout          int64
Dalc           int64
Walc           int64
health         int64
absences       int64
G1             int64
G2             int64
G3             int64
dtype: object

### Step 2.b: Check for null (missing) values
- Data often is missing entries - there can be many reasons for this
- We need to deal with that (will do later in course)
- Use ```.isnull().any()```

In [7]:
data.isnull().any()

school        False
sex           False
age           False
address       False
famsize       False
Pstatus       False
Medu          False
Fedu          False
Mjob          False
Fjob          False
reason        False
guardian      False
traveltime    False
studytime     False
failures      False
schoolsup     False
famsup        False
paid          False
activities    False
nursery       False
higher        False
internet      False
romantic      False
famrel        False
freetime      False
goout         False
Dalc          False
Walc          False
health        False
absences      False
G1            False
G2            False
G3            False
dtype: bool

## Step 3: Analyze
- Feature selection
- Model selection
- Analyze data

### Description
- Want to find 3 features to use in our report
- The 3 features should be selected based on
    - Actionable insights
    - Convey credibility in report
    - What is realistic within possibilities (including a budget)

### Note
- This step is where you can explore
- You know how to use the following:
    - **corr()** to find see correlation
    - **groupby()** with **mean()**, **count()**, or **std()**
- This should be used for step 4: Report

### Step 3.a: Investigate correlation
- Correlation is an easy measure to find insights that are actionable.
- Use **corr()** and only show **G3**, as that is the row we are interested in.
    - Notice: **G1** and **G2** are highly correlated, but they are not intented to be used

In [13]:
data.corr()['G3']

age          -0.161579
Medu          0.217147
Fedu          0.152457
traveltime   -0.117142
studytime     0.097820
failures     -0.360415
famrel        0.051363
freetime      0.011307
goout        -0.132791
Dalc         -0.054660
Walc         -0.051939
health       -0.061335
absences      0.034247
G1            0.801468
G2            0.904868
G3            1.000000
Name: G3, dtype: float64

### Step 3.b: Get the Feature names
- This step can help you understand features better.
- All the features are availbale witb **.columns** applied on the **DataFrame**

In [14]:
data.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'G1', 'G2', 'G3'],
      dtype='object')

### Step 3.c: Investigate features
**Repeat this step** (possibly for all features)
- Select a features
- Calculate the **groupby(...)** **mean()** on **G3**
    - HINT: This was done in the lesson
- Calculate the **groupby(...)** **count()** on **G3**
- Calculate the **groupby(...)** **std()** on **G3**

In [16]:
data.groupby('higher').mean()['G3']

higher
no      6.800
yes    10.608
Name: G3, dtype: float64

In [17]:
data.groupby('higher').count()['G3']

higher
no      20
yes    375
Name: G3, dtype: int64

In [18]:
data.groupby('higher').std()['G3']

higher
no     4.829732
yes    4.493422
Name: G3, dtype: float64

In [19]:
data.groupby('address').mean()['G3']

address
R     9.511364
U    10.674267
Name: G3, dtype: float64

In [48]:
data.groupby('address').count()['G3']

address
R     88
U    307
Name: G3, dtype: int64

In [21]:
data.groupby('address').std()['G3']

address
R    4.556149
U    4.563075
Name: G3, dtype: float64

In [28]:
data.groupby('traveltime').mean()['G3']

traveltime
1    10.782101
2     9.906542
3     9.260870
4     8.750000
Name: G3, dtype: float64

In [29]:
data.groupby('traveltime').count()['G3']

traveltime
1    257
2    107
3     23
4      8
Name: G3, dtype: int64

In [30]:
data.groupby('address').mean()['traveltime']

address
R    1.875000
U    1.325733
Name: traveltime, dtype: float64

In [31]:
data.groupby('address').std()['traveltime']

address
R    0.881646
U    0.581374
Name: traveltime, dtype: float64

In [39]:
data.groupby('traveltime').mean()['failures']

traveltime
1    0.299611
2    0.355140
3    0.478261
4    0.750000
Name: failures, dtype: float64

In [44]:
data.groupby('address').mean()['absences']

address
R    6.125000
U    5.589577
Name: absences, dtype: float64

In [52]:
data.groupby('higher').mean()['traveltime']

higher
no     1.700000
yes    1.434667
Name: traveltime, dtype: float64

In [54]:
data.groupby('higher').std()['traveltime']

higher
no     0.864505
yes    0.686285
Name: traveltime, dtype: float64

### Step 3.d Select 3 features
- Decide on 3 features to use in the report
- The decision should be based on
    - Actionable insights
    - Convey credibility in report
    - What is realistic within possibilities (including a budget)

In [77]:
data.groupby('higher').mean()['G3']

higher
no      6.800
yes    10.608
Name: G3, dtype: float64

In [78]:
data.groupby('higher').std()['G3']

higher
no     4.829732
yes    4.493422
Name: G3, dtype: float64

In [63]:
#Rural students
data.groupby('address').count()['G3']['R']/len(data)*100

22.278481012658226

In [64]:
data.groupby('address').mean()['traveltime']

address
R    1.875000
U    1.325733
Name: traveltime, dtype: float64

In [55]:
data.groupby('address').mean()['G3']

address
R     9.511364
U    10.674267
Name: G3, dtype: float64

In [65]:
data.groupby('traveltime').mean()['G3']

traveltime
1    10.782101
2     9.906542
3     9.260870
4     8.750000
Name: G3, dtype: float64

In [66]:
data.groupby('address').mean()['absences']

address
R    6.125000
U    5.589577
Name: absences, dtype: float64

## Step 4: Report
- Present findings
- Visualize results
- Credibility counts

### Description
- With the 3 features from step 3 create a presentation
- As we have not learned visualization yet, keep it simple
- Remember, that credibility counts

### Notice
- At this stage it is not supposed to be perfect.
- Present the findings here in the Notebook

There are two branches that we found.

The first one is the motivation, when someone asks students if they want to improve their grades, it becomes a flag to watch because the grades of those who don´t want to improve go between 2 - 11.6. Instead of students who want to improve, their grades are between 6.2 - 15.

The other branch is their addresses, rural students take more time to go to school, the results show that while more travel time, students get a worse grade; besides, absences are more common for rural students that represent 22%.

## Step 5: Actions
- Use insights
- Measure impact
- Main goal

### Description
- What actions should the schools take?
- How can they evaluate the impact?
- Remember, this is the main goal.

Motivation is one of the most important feelings that students could or not have. The student can still go to school and do their activities, but lack of motivation could affect the whole future of the student. Has been demostrated that unmotivated students have worse grades than motivaded ones. 

In our case, the grades of those who don´t want to improve go between 2 - 11.6, instead of students who want to improve, their grades are between 6.2 - 15.

One thing that schools can do to improve motivation rate could be have more activities where the student could be identified whit their personality or their study method.

Other fact to considerate is the distance between schools and student´s address. Rural students take more time to go to school, the results show that while more travel time, students get a worse grade; besides, absences are more common for rural students.

Is not much that schools can do in this cases, because principals can´t just bluid more schools, that depends on the state.